### ISSUE TITLE
Phase 41.3 — AdversarialTrainer: PGD-AT, TRADES, AWP & Curriculum Training

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#827]
# Phase 41.3 — AdversarialTrainer

## Overview

The **AdversarialTrainer** implements state-of-the-art adversarial training methods that harden models against adversarial attacks during the training process itself. It implements PGD-AT (Madry et al., 2018), TRADES (Zhang et al., 2019), AWP (Wu et al., 2020), and curriculum adversarial training — the most effective empirical defenses known.

## Architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                      AdversarialTrainer                          │
│                                                                  │
│  ┌──────────────────┐  ┌──────────────────┐  ┌──────────────┐  │
│  │  Training Mode    │  │  Loss Functions   │  │  Curriculum  │  │
│  │  Selector         │  │                   │  │  Scheduler   │  │
│  │                   │  │  PGD-AT Loss      │  │              │  │
│  │  Standard AT      │  │  TRADES Loss      │  │  ε-schedule  │  │
│  │  TRADES           │  │  MART Loss        │  │  step-sched  │  │
│  │  AWP              │  │  AWP Regularizer  │  │  mix ratio   │  │
│  │  TRADES + AWP     │  │  KL-divergence    │  │  warm-up     │  │
│  │  Free AT          │  │                   │  │              │  │
│  │  YOPO             │  │                   │  │              │  │
│  └────────┬─────────┘  └────────┬──────────┘  └──────┬───────┘  │
│           │                      │                    │          │
│  ┌────────▼──────────────────────▼────────────────────▼───────┐  │
│  │              Adversarial Training Loop                      │  │
│  │                                                             │  │
│  │  for each epoch:                                            │  │
│  │    for each batch (x, y):                                   │  │
│  │      1. Generate adversarial examples: x_adv = Attack(x,y) │  │
│  │      2. Compute robust loss: L(f(x_adv), y) [+ reg]        │  │
│  │      3. [AWP] Perturb weights → recompute loss → restore   │  │
│  │      4. Update model: θ ← θ - η·∇L                         │  │
│  │      5. Evaluate clean + robust accuracy                    │  │
│  └─────────────────────────────────────────────────────────────┘  │
│                                                                  │
│  ┌─────────────────────────────────────────────────────────────┐  │
│  │              Checkpoint & Evaluation Manager                 │  │
│  │  Best robust acc • Learning curves • Overfitting detection  │  │
│  └─────────────────────────────────────────────────────────────┘  │
└─────────────────────────────────────────────────────────────────┘
```

## TypeScript Interface

```typescript
import { Tensor, Model, Optimizer, DataLoader } from '@asi-build/core';

type TrainingMethod = 'pgd_at' | 'trades' | 'awp' | 'trades_awp' | 'free_at' | 'yopo' | 'mart';

interface AdversarialTrainingConfig {
  method: TrainingMethod;
  epsilon: number;                // perturbation budget
  norm: 'Linf' | 'L2';
  attackSteps: number;            // inner maximization steps
  attackStepSize: number;
  randomStart: boolean;
  
  // TRADES-specific
  tradesBeta?: number;            // λ in TRADES (natural vs robust tradeoff)
  
  // AWP-specific
  awpGamma?: number;              // weight perturbation magnitude
  awpWarmup?: number;             // epochs before enabling AWP
  
  // Curriculum
  curriculumEpsilon?: boolean;    // gradually increase epsilon
  epsilonStart?: number;
  epsilonEnd?: number;
  epsilonWarmupEpochs?: number;
  
  // Training
  epochs: number;
  learningRate: number;
  lrSchedule: 'cosine' | 'multistep' | 'cyclic';
  weightDecay: number;
  batchSize: number;
  
  // Early stopping
  earlyStoppingPatience?: number;
  earlyStoppingMetric?: 'robust_accuracy' | 'combined';
}

interface TrainingMetrics {
  epoch: number;
  trainLoss: number;
  cleanAccuracy: number;
  robustAccuracy: number;        // accuracy under PGD attack
  bestRobustAccuracy: number;
  learningRate: number;
  currentEpsilon: number;
  awpActive: boolean;
  timePerEpoch: number;
}

interface AdversarialTrainer {
  // Core training methods
  trainPGDAT(model: Model, trainLoader: DataLoader, config: AdversarialTrainingConfig): AsyncIterable<TrainingMetrics>;
  trainTRADES(model: Model, trainLoader: DataLoader, config: AdversarialTrainingConfig): AsyncIterable<TrainingMetrics>;
  trainAWP(model: Model, trainLoader: DataLoader, config: AdversarialTrainingConfig): AsyncIterable<TrainingMetrics>;
  
  // Unified interface
  train(model: Model, trainLoader: DataLoader, config: AdversarialTrainingConfig): AsyncIterable<TrainingMetrics>;
  
  // Evaluation
  evaluateRobustness(model: Model, testLoader: DataLoader, epsilon: number): Promise<{clean: number, robust: number}>;
  
  // Checkpoint management
  saveCheckpoint(model: Model, metrics: TrainingMetrics, path: string): Promise<void>;
  loadBestCheckpoint(path: string): Promise<{model: Model, metrics: TrainingMetrics}>;
  
  // Robustness-accuracy tradeoff analysis
  tradeoffCurve(model: Model, testLoader: DataLoader, epsilons: number[]): Promise<Map<number, {clean: number, robust: number}>>;
}
```

## TRADES Loss Implementation

```python
class TRADESLoss:
    """Zhang et al. (2019) — Theoretically Principled Trade-off."""
    
    def __init__(self, beta: float = 6.0, epsilon: float = 8/255,
                 step_size: float = 2/255, num_steps: int = 10):
        self.beta = beta
        self.epsilon = epsilon
        self.step_size = step_size
        self.num_steps = num_steps
    
    def __call__(self, model, x_natural, y):
        # Natural loss
        logits_natural = model(x_natural)
        loss_natural = F.cross_entropy(logits_natural, y)
        
        # Generate adversarial examples (maximizing KL divergence)
        model.eval()
        x_adv = x_natural.detach() + 0.001 * torch.randn_like(x_natural)
        
        for _ in range(self.num_steps):
            x_adv.requires_grad_(True)
            with torch.enable_grad():
                loss_kl = F.kl_div(
                    F.log_softmax(model(x_adv), dim=1),
                    F.softmax(logits_natural.detach(), dim=1),
                    reduction='batchmean'
                )
            grad = torch.autograd.grad(loss_kl, [x_adv])[0]
            x_adv = x_adv.detach() + self.step_size * grad.sign()
            x_adv = torch.min(torch.max(x_adv, x_natural - self.epsilon), 
                              x_natural + self.epsilon)
            x_adv = torch.clamp(x_adv, 0.0, 1.0)
        
        model.train()
        
        # TRADES loss = CE(f(x), y) + β · KL(f(x) || f(x_adv))
        logits_adv = model(x_adv)
        loss_robust = F.kl_div(
            F.log_softmax(logits_adv, dim=1),
            F.softmax(logits_natural.detach(), dim=1),
            reduction='batchmean'
        )
        
        return loss_natural + self.beta * loss_robust
```

## AWP Implementation

```python
class AWP:
    """Wu et al. (2020) — Adversarial Weight Perturbation."""
    
    def __init__(self, model, gamma: float = 0.01):
        self.model = model
        self.gamma = gamma
        self.param_backup = {}
        self.grad_backup = {}
    
    def perturb(self):
        """Add worst-case weight perturbation."""
        self._save_params()
        for name, param in self.model.named_parameters():
            if param.requires_grad and param.grad is not None:
                norm = torch.norm(param.grad)
                if norm != 0:
                    perturbation = self.gamma * param.grad / norm
                    param.data.add_(perturbation)
    
    def restore(self):
        """Restore original weights."""
        for name, param in self.model.named_parameters():
            if name in self.param_backup:
                param.data = self.param_backup[name]
        self.param_backup = {}
```

## Configuration

```yaml
adversarial_trainer:
  method: "trades_awp"
  
  pgd_at:
    epsilon: 0.031373             # 8/255
    attack_steps: 10
    step_size: 0.007843           # 2/255
    random_start: true
    
  trades:
    beta: 6.0                    # robustness weight
    epsilon: 0.031373
    attack_steps: 10
    step_size: 0.007843
    
  awp:
    gamma: 0.01                  # weight perturbation magnitude
    warmup_epochs: 10            # start AWP after 10 epochs
    
  curriculum:
    enabled: true
    epsilon_start: 0.0
    epsilon_end: 0.031373
    warmup_epochs: 5
    
  training:
    epochs: 200
    batch_size: 128
    learning_rate: 0.1
    lr_schedule: "multistep"
    lr_milestones: [100, 150]
    lr_gamma: 0.1
    weight_decay: 0.0005
    momentum: 0.9
    
  evaluation:
    eval_frequency: 5            # evaluate every N epochs
    pgd_eval_steps: 20
    save_best: true
    early_stopping_patience: 30
```

## Testing Strategy

```typescript
describe('AdversarialTrainer', () => {
  describe('PGD-AT', () => {
    it('should improve robust accuracy over standard training', async () => {
      const standardModel = await trainStandard(model, trainLoader, 10);
      const robustModel = await trainer.trainPGDAT(model, trainLoader, { ...config, epochs: 10 });
      
      const standardRobust = await evaluateRobustness(standardModel, testLoader, 0.031);
      const atRobust = await evaluateRobustness(robustModel, testLoader, 0.031);
      
      expect(atRobust.robust).toBeGreaterThan(standardRobust.robust);
    });
  });
  
  describe('TRADES', () => {
    it('should achieve better robustness-accuracy tradeoff than PGD-AT', async () => {
      const tradesMetrics = await lastOf(trainer.trainTRADES(model, trainLoader, tradesConfig));
      expect(tradesMetrics.cleanAccuracy + tradesMetrics.robustAccuracy).toBeGreaterThan(threshold);
    });
    
    it('should respect beta parameter for tradeoff control', async () => {
      const lowBeta = await lastOf(trainer.trainTRADES(model, trainLoader, { ...config, tradesBeta: 1.0 }));
      const highBeta = await lastOf(trainer.trainTRADES(model, trainLoader, { ...config, tradesBeta: 10.0 }));
      expect(highBeta.robustAccuracy).toBeGreaterThan(lowBeta.robustAccuracy);
      expect(lowBeta.cleanAccuracy).toBeGreaterThan(highBeta.cleanAccuracy);
    });
  });
  
  describe('Curriculum', () => {
    it('should gradually increase epsilon during warmup', async () => {
      const metrics: TrainingMetrics[] = [];
      for await (const m of trainer.train(model, trainLoader, { ...config, curriculumEpsilon: true })) {
        metrics.push(m);
      }
      expect(metrics[0].currentEpsilon).toBeLessThan(metrics[metrics.length - 1].currentEpsilon);
    });
  });
});
```

## Acceptance Criteria

- [ ] PGD-AT achieves ≥45% robust accuracy on CIFAR-10 at ε=8/255
- [ ] TRADES achieves ≥48% robust accuracy with ≥80% clean accuracy
- [ ] AWP improves robust generalization gap by ≥2% over base TRADES
- [ ] Curriculum training converges faster than constant-epsilon training
- [ ] Checkpoint system saves/restores best robust model correctly
- [ ] Training loop supports gradient accumulation for large batch sizes
- [ ] Metrics logged per-epoch with clean accuracy, robust accuracy, loss components
- [ ] Early stopping triggers on robust accuracy plateau

## References

- Madry et al. (2018) — Towards Deep Learning Models Resistant to Adversarial Attacks
- Zhang et al. (2019) — Theoretically Principled Trade-off (TRADES)
- Wu et al. (2020) — Adversarial Weight Perturbation (AWP)
- Rice et al. (2020) — Overfitting in Adversarially Robust Deep Learning
- Pang et al. (2021) — Bag of Tricks for Adversarial Training

---
*Sub-phase of [Phase 41 — Adversarial Robustness & Security Intelligence](../discussions/824)*
